packages/gestalt/src/AvatarGroup.js

```diff
@@ -131,58 +131,48 @@ const AvatarGroupWithForwardRef: AbstractComponent<Props, UnionRefs> = forwardRe
       (showCollaboratorsCount ? 1 : 0) +
       (showAddCollaboratorsButton ? 1 : 0);
 
-    let collaboratorStack: $ReadOnlyArray<ReactNode> = displayedCollaborators.map(
-      ({ src, name }, index) => (
-        <CollaboratorAvatar
-          hovered={hovered}
-          index={index}
-          // eslint-disable-next-line react/no-array-index-key
-          key={`collaboratorStack-${name}-${index}`}
-          name={name}
-          pileCount={pileCount}
-          size={size}
-          src={src || ''}
-        />
-      ),
+    const collaboratorStack = (
+      <Fragment>
+        {displayedCollaborators.map(({ src, name }, index) => (
+          <CollaboratorAvatar
+            hovered={hovered}
+            index={index}
+            key={`collaboratorStack-${name}-${index}`} // eslint-disable-line react/no-array-index-key
+            name={name}
+            pileCount={pileCount}
+            size={size}
+            src={src || ''}
+          />
+        ))}
+        {showCollaboratorsCount && (
+          <CollaboratorsCount
+            count={collaborators.length - 2}
+            showAddCollaboratorsButton={showAddCollaboratorsButton}
+            hovered={hovered}
+            key={`collaboratorStack-count-${collaborators.length}`}
+            pileCount={pileCount}
+            size={size}
+          />
+        )}
+        {showAddCollaboratorsButton && (
+          <AddCollaboratorsButton
+            hovered={hovered}
+            key={`collaboratorStack-addButton-${collaborators.length}`}
+            pileCount={pileCount}
+            size={size}
+          />
+        )}
+      </Fragment>
     );
 
-    if (showCollaboratorsCount) {
-      collaboratorStack = [
-        ...collaboratorStack,
-        <CollaboratorsCount
-          count={collaborators.length - 2}
-          showAddCollaboratorsButton={showAddCollaboratorsButton}
-          hovered={hovered}
-          key={`collaboratorStack-count-${collaborators.length}`}
-          pileCount={pileCount}
-          size={size}
-        />,
-      ];
-    }
-
-    if (showAddCollaboratorsButton) {
-      collaboratorStack = [
-        ...collaboratorStack,
-        <AddCollaboratorsButton
-          hovered={hovered}
-          key={`collaboratorStack-addButton-${collaborators.length}`}
-          pileCount={pileCount}
-          size={size}
-        />,
-      ];
-    }
-
-    const AvatarGroupStack = useCallback(
```

Comment:
It's so curious that this was actually causing the bug, not the array operations above!

Reply:
Did you figure out which of `accessibilityLabel, collaboratorStack, isDisplayOnly, isFitSize` was changing?

Reply:
It was `collaboratorStack` changing on every render. It's kinda weird that even when I wrapped it with useMemo, the bug still persisted.

Reply:
Did you wrap it in `useMemo` before or after you refactored to define it using a `const` and fragments?

Reply:
I tried useMemo before refactoring. I think the root of the problem is the `.slice` method that is used to create `displayedCollaborators`. Because it creates a new array instance. The dependecy list of the hooks are compared by reference.
But how does it work without any hooks? The idea is to convert the array data into components. React does quite a good job of "diff"ing component trees (than the list of dependencies) and prevents unecessary renders.

Reply:
Ah, yep. That would be expected then. We were creating new array references all over the place before.

And correct, while the new logic does create a new variable full of components on each render, React is smart enough to not update the DOM if nothing in those components is changing.

Nice fix!